ENH: Add xvals param to lowess smoother

tgbrooks opened the pull request:
xvals allows evaluation of the lowess smoother at
x-values other than those used in the smoothing, for example
fitting from some points and using that to 'predict' others

- [x] closes #6543
- [x] tests added / passed. 
- [x] code/documentation is well formatted.  
- [x] properly formatted commit message. See 
      [NumPy's guide](https://docs.scipy.org/doc/numpy-1.15.1/dev/gitwash/development_workflow.html#writing-the-commit-message). 

<details>


**Notes**:

* It is essential that you add a test when making code changes. Tests are not 
  needed for doc changes.
* When adding a new function, test values should usually be verified in another package (e.g., R/SAS/Stata).
* When fixing a bug, you must add a test that would produce the bug in master and
  then show that it is fixed with the new code.
* New code additions must be well formatted. Changes should pass flake8. If on Linux or OSX, you can
  verify you changes are well formatted by running 
  ```
  git diff upstream/master -u -- "*.py" | flake8 --diff --isolated
  ```
  assuming `flake8` is installed. This command is also available on Windows 
  using the Windows System for Linux once `flake8` is installed in the 
  local Linux environment. While passing this test is not required, it is good practice and it help 
  improve code quality in `statsmodels`.
* Docstring additions must render correctly, including escapes and LaTeX.

</details>

bashtage commented on statsmodels/nonparametric/_smoothers_lowess.pyx:
```diff
@@ -24,12 +24,16 @@ cdef inline double fmax(double x, double y): return x if x >= y else y
 
 DTYPE = np.double
 ctypedef np.double_t DTYPE_t
+cdef double NAN = float("NaN")
```
You could just use `cdef double NAN = 0.0 / 0.0` or 

```
from cython.math cimport nan
cdef double NAN = nan("")
```

bashtage commented on statsmodels/nonparametric/_smoothers_lowess.pyx:
```diff
@@ -42,6 +46,8 @@ def lowess(np.ndarray[DTYPE_t, ndim = 1] endog,
         The y-values of the observed points
     exog : 1-D numpy array
         The x-values of the observed points. exog has to be increasing.
+    weights : 1-D numpy array
```
weights or resid_weights?

bashtage commented on statsmodels/nonparametric/_smoothers_lowess.pyx:
```diff
@@ -51,13 +57,19 @@ def lowess(np.ndarray[DTYPE_t, ndim = 1] endog,
         A numpy array with two columns. The first column
         is the sorted x values and the second column the
         associated estimated y-values.
+    resid_weights: numpy array
```
Are there different from weights?  Also (throughout ideally, return type should be `ndarray` as in `resid_weights: ndarray`.

tgbrooks replied:
'weights' are the data point weights for regressing at a specific point, while `resid_weights` is used for the overall down-weighting of outliers produced from multiple iterations and so don't depend upon the x values.

bashtage commented on statsmodels/nonparametric/tests/test_lowess.py:
```diff
@@ -120,6 +120,16 @@ def test_options(self):
         assert_almost_equal(actual_lowess, actual_lowess1, decimal=13)
         assert_(actual_lowess.dtype is np.dtype(float))
 
+        # Test specifying xvals explicitly
```
Could these be added as new tests of the feature, rather than to an existing test?

bashtage commented on statsmodels/nonparametric/smoothers_lowess.py:
```diff
@@ -32,10 +32,14 @@ def lowess(endog, exog, frac=2.0/3.0, it=3, delta=0.0, is_sorted=False,
     delta : float
         Distance within which to use linear-interpolation
         instead of weighted regression.
+    xvals: 1-D numpy array
```
`xvals : ndarray, 1d`

bashtage commented on statsmodels/nonparametric/smoothers_lowess.py:
```diff
@@ -10,7 +10,7 @@
 from ._smoothers_lowess import lowess as _lowess
 
-def lowess(endog, exog, frac=2.0/3.0, it=3, delta=0.0, is_sorted=False,
+def lowess(endog, exog, frac=2.0/3.0, it=3, delta=0.0, xvals=None, is_sorted=False,
```
The new optional should, unfortunately, go at the end.  I cold also live with making everything after (and including) `it=3` keyword only, so that `frac=2.0/3.0, *, it=3,...`.

bashtage replied:
This breaks code but in a way that is loud and obvious.  Leaving it were it is could like to hard to trace breaks.

josef-pkt commented:
I think the name in the lowess function should be `exog_predict` instead of xvals.
In internal code `xvals` is ok.

tgbrooks commented:
I've now made all the changes I wanted to make. Thanks for the feedback and let me know if anything else should be changed or added.

bashtage commented:
I checked coverage in the pyx files -- all of the new code is covered by your tests.

bashtage commented:
Any chance you could add the use of this to a notebook (or create a new one) to advertise this feature?  Ideally to produce confidence intervals as you intended in the first place.

tgbrooks commented:
That sounds like a good idea. It looks like the repo has .ipynb and .py files in examples/ but that the .py autogenerated from the .ipynb. How will I autogenerate it? Just export manually from jupyter?

josef-pkt commented:
Only commit notebooks without the output to save on repo size, i.e. clear output before committing

The py files are generated by manual export, so it does not always happen.
One advantage of also having the py-files is that they provides less noise changesets if we need to check the history (which doesn't happen very often for notebooks).

bashtage commented:
There is a script in tools that generates the py files.

On Fri, Feb 28, 2020, 19:28 Thomas Brooks <notifications@github.com> wrote:

> That sounds like a good idea. It looks like the repo has .ipynb and .py
> files in examples/ but that the .py autogenerated from the .ipynb. How will
> I autogenerate it? Just export manually from jupyter?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/statsmodels/statsmodels/pull/6554?email_source=notifications&email_token=ABKTSRLFPBT2OU2F5RBHCSDRFFQXDA5CNFSM4K4MCOU2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENJ3K2Y#issuecomment-592688491>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKTSRL3KCFJXP4NDCYNOFTRFFQXDANCNFSM4K4MCOUQ>
> .
>

tgbrooks commented:
I added the ipython notebook with a simple example. As well, I changed the `delta` behavior I mentioned above: now `delta` is not used on the `exog_predict` points, only on the `exog` and `endog` points. This means that the value of the fit at a point of `exog_predict` does not depend upon the other values in `exog_predict`. I think this reflects the most typical use-case best.

Let me know if any other changes are needed.

bashtage commented:
Looks great. Many thanks; this will be in the next release.
